### DISCUSSION TITLE
Multi-stage pipeline: Why wait for two groups to complete at the beginning?

### DISCUSSION THREAD

**Discussion by masahi** [NVIDIA/cutlass#500] [Q&A] [answered]
In https://github.com/NVIDIA/cutlass/blob/12f4108ac2233022b92f0e3533c23bed399fcf45/include/cutlass/gemm/threadblock/mma_multistage.h#L433, we wait for two oldest stages of async copy groups to finish. Why do we need to wait for two, instead of one? 

If my read of the innermost `warp_mma_k` loop is correct, we only consume one-stage worth of async copy result in the inner most loop before we commit another group and wait for the next oldest group to be ready at https://github.com/NVIDIA/cutlass/blob/12f4108ac2233022b92f0e3533c23bed399fcf45/include/cutlass/gemm/threadblock/mma_multistage.h#L555-L563. 

So I want to know why we keep two stages of copy results always ready in the innermost loop.

**Comment by masahi** [comment, marked as answer]
oh maybe I got it. At the last `warp_mma_k` iteration of a given stage, we want to start loading from the next stage copy results:
https://github.com/NVIDIA/cutlass/blob/12f4108ac2233022b92f0e3533c23bed399fcf45/include/cutlass/gemm/threadblock/mma_multistage.h#L500-L501

so yeah, I can see why we need two stages of copy results always ready.

**Comment by hwu36** [comment]
` cutlass::arch::cp_async_wait<Base::kStages - 2>(); ` means 2 stages are still on the fly and the rest are committed.  Check this [comment](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/arch/memory_sm80.h#L401).

> we only consume one-stage worth of async copy result in the inner most loop before we commit another group and wait for the next oldest group to be ready

Correct.

> At the last warp_mma_k iteration of a given stage, we want to start loading from the next stage copy results:

Yes, we do this.

**Reply by masahi** [reply]
> means 2 stages are still on the fly and the rest are committed

Isn't it the opposite? I thought it means 2 stages are completed and `Base::kStages - 2` stages are still in flight. I'm confused with the meaning of "committed" in the comment:

https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/arch/memory_sm80.h#L401

If it said "Blocks until all but N previous cp.async.commit_group operations have **completed**", that agrees with my understanding of the instruction, according to the PTX doc https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-wait-group

**Comment by hwu36** [comment]
> sn't it the opposite? I thought it means 2 stages are completed and Base::kStages - 2 stages are still in flight. 

You are correct.  I misspoke.

**Comment by TigerYang414** [comment]
Is this multi-stage discussed here related to stages argument used to init tile description?
If it is, the stages here means the number of stages used to copy memory to shared memory?
But for op code of simt, I only see 2 stages used, is there any limit for this?

**Comment by hwu36** [comment]
> Is this multi-stage discussed here related to stages argument used to init tile description?

yes

> If it is, the stages here means the number of stages used to copy memory to shared memory?

yes

> But for op code of simt, I only see 2 stages used, is there any limit for this?

sgemm on sm80 can use multiple stages.

**Reply by TigerYang414** [reply]
So what determines stages we can use for different sms? Is there any doc or best practice guidelines?

**Comment by hwu36** [comment]
multistage is only useful for >=  sm80.  you can profile to choose the best stage for you.  you can start from those used by cutlass profiler.
